pkg/util/timeout.go

```diff
@@ -14,29 +14,38 @@ type TimeoutError struct {
 // Error implements the Go error interface.
 func (t *TimeoutError) Error() string {
 	if len(t.message) > 0 {
-		return fmt.Sprintf(t.message, t.after)
+		return fmt.Sprintf("%s timed out after %v", t.message, t.after)
 	}
-	return fmt.Sprintf("calling the function timeout after %v", t.after)
+	return fmt.Sprintf("function timed out after %v", t.after)
 }
 
-// TimeoutAfter executes the provide function and return the TimeoutError in
-// case when the execution time of the provided function is bigger than provided
-// time duration.
-func TimeoutAfter(t time.Duration, errorMsg string, fn func() error) error {
+// TimeoutAfter executes the provided function and returns TimeoutError in the
+// case that the execution time of the function exceeded the provided duration.
+// The provided function is passed the timer in case it wishes to reset it.  If
+// so, the following pattern must be used:
+// if !timer.Stop() {
+//   return &TimeoutError{}
+// }
+// timer.Reset(timeout)
+func TimeoutAfter(t time.Duration, errorMsg string, f func(*time.Timer) error) error {
 	c := make(chan error, 1)
+	timer := time.NewTimer(t)
 	go func() {
-		defer close(c)
-		c <- fn()
+		err := f(timer)
+		if !IsTimeoutError(err) {
+			c <- err
+		}
```

Comment:
A caller of TimeoutAfter would expect a TimeoutError with its fields populated, and the TimeoutAfter function has the information necessary to do this.  However the TimeoutAfter function does not currently send this information into f().  f() needs a way to signal to TimeoutAfter "I realise I've timed out, I'm giving up and exiting politely".  Within its signature (returning error) the only way of doing this is by return of a special error type.  I've reused TimeoutError for this purpose.  Thus the TimeoutError from f() needs to be found and quietly discarded; the TimeoutError that TimeoutAfter sends will be correctly populated.

Alternatives here would be:
- add another error type for f() to return instead of TimeoutError (don't overload usage of TimeoutError)
- pass errorMsg into f() so any TimeoutError it may return is populated correctly and bubbles up either way
- change the signature of f() so it sends to a channel instead of returning an error, thus f() can decide not to send before it returns